update_films_data: Обновление видеотеки Smarty

BergaBruh opened the pull request:
* Добавлена информация о скрипте
* Добавлен скрипт на обновление видеотеки (требуется `kinopoisk_id` в Smarty)

BergaBruh committed:
update_films_data: обновление данных у фильма

Примечание: требуется kinopoisk_id в Smarty, чтобы можно было обновить данные

BergaBruh committed:
Update README.md

ArthurO-micimp commented on content-api/python/update_films_data.py:
```diff
@@ -0,0 +1,50 @@
+content_api_key = 'top secret'  # ContentAPI ключ. Нужно для обновления данных о фильмах
+
+
+def get_from_smarty_films_in_the_form_json_list(page):
```
Функция, название которой начинается на get должна что-то возвращать, поэтому надо переименовать.

ArthurO-micimp commented on content-api/python/update_films_data.py:
```diff
@@ -0,0 +1,50 @@
+
+
+def get_from_smarty_films_in_the_form_json_list(page):
+    limit_film = 100 # Количество фильмов/Сериалов на одну страницу
```
Это можно вынести в объявление функции.

ArthurO-micimp commented on content-api/python/update_films_data.py:
```diff
@@ -0,0 +1,50 @@
+    limit_film = 100 # Количество фильмов/Сериалов на одну страницу
+    number = 0
+
+    get_api_response = api.video_list(limit=limit_film, page=page)
```
api берется из глобального пространства - так делать не стоит. Надо вынести api в объявление функции.

ArthurO-micimp commented on content-api/python/update_films_data.py:
```diff
@@ -0,0 +1,50 @@
+    get_from_smarty_films_data(videos_info, video_count, number, page, limit_film)
+
+
+def get_from_smarty_films_data(videos_info, video_count, number, page, limit_film):
```
Аналогично, название функции не совсем отражает то, что она делает.

ArthurO-micimp commented on content-api/python/update_films_data.py:
```diff
@@ -0,0 +1,50 @@
+
+
+def get_from_smarty_films_data(videos_info, video_count, number, page, limit_film):
+    while video_count > page:
```
Сравнение количества видео с номером страницы не выглядит корректно. Либо тут логическая ошибка, либо неправильное название для переменных.

ArthurO-micimp commented on content-api/python/update_films_data.py:
```diff
@@ -0,0 +1,50 @@
+            try:
+                api.video_modify(id=id_number, load_meta=1) # Обновление данных фильма (требуется вбитый kinopoisk_id в Smarty)
+                logging.info(
+                    f"Для фильма/сериала с ID {id_number} была обновлена информация через Kinopoisk Unofficial "
+                    f"API и добавлена в Smarty.")
```
Лучше просто написать что была загружена метаинформация. Метаинформация может подтягиваться и из других источников тоже.

ArthurO-micimp commented on content-api/python/update_films_data.py:
```diff
@@ -0,0 +1,50 @@
+            number += 1
+            if number == limit_film:
+                page += 1
+                get_from_smarty_films_in_the_form_json_list(page)
```
Рекурсию лучше избегать, у нее есть свойство занимать весь стэк и заканчивать на этом выполнение.
Лучше всего в данном случае было бы сделать обновление данных через цикл. Для этого надо сначала сделать запрос video_list с параметром limit=0, чтобы получить сколько всего фильмов. После этого циклом через каждые N фильмов получать их и обновлять:
```
response = api.video_list(limit=0,page=1)
count = response['count']
limit = 100
for i in range(1, int(count/limit)):
    response = api.video_list(limit=limit,page=i)
    videos = response['videos']
    update_videos(api, videos)
```

ArthurO-micimp reviewed (changes requested):
Можно переделать по аналогии с sync_content, сделать отдельный класс, которому при инициализации надо передать api.

А еще не хватает тайпхинтов в функциях.

BergaBruh committed:
update_films_data: переписывание кода

Код был переписан по аналогии с sync_content

BergaBruh commented:
Скрипт был переписан по аналогии с sync_content. Требуется перепроверка кода.

ArthurO-micimp commented on content-api/python/update_films_data.py:
```diff
@@ -0,0 +1,69 @@
+        return pages
+
+    def update_video(self, video_files: int):
+        try:
+            api.video_modify(id=video_files, load_meta=1)
+        except URLError or HTTPError:
+            ParseError()
```
Чтобы бросить исключение, надо писать `raise ParseError`. Да и название у исключения некорректное - исключение возникает из-за подключения, а не парсинга. Впрочем, здесь вообще нет смысла делать отдельное исключение, так как обрабатывать каждое исключение в данном случае довольно трудно, поэтому при первой же ошибке достаточно вывести ошибку и прекратить выполнение.

ArthurO-micimp commented on content-api/python/update_films_data.py:
```diff
@@ -0,0 +1,69 @@
+                self.update_video(video_files)
+                print(f'Информация о фильме с ID = {video_files} обновлена успешно')
+                logging.info(f'Информация о фильме с ID = {video_files} обновлена успешно')
+            except ContentAPIException as error:
```
Здесь отловится только одна ошибка: ContentAPIException, все остальные вывалятся во внешку и, скорее всего, не будут обработаны. В данном случае можно ловить вообще все исключения: `except Exception as e:`

ArthurO-micimp commented on content-api/python/update_films_data.py:
```diff
@@ -0,0 +1,69 @@
+                              f' Сообщение об ошибки: {repr(error)}')
+                break
+
+    def load_data(self):
```
Можно назвать метод `run`, тогда будет понятно, что он запустит процесс обновления данных.

ArthurO-micimp commented on content-api/python/update_films_data.py:
```diff
@@ -0,0 +1,69 @@
+
+
+class UpdateVideoList(object):
+    def __init__(self, api, videos_per_page):
```
Вот тут точно стоит добавить тайпхинты, так как этот класс, по идее, вызывается сторонними разработчиками, поэтому им будет лучше, если они будут знать какие типы нужно передавать в аргументы:
`api: SmartyContentAPI`
videos_per_page сам по себе к классу не относится, его можно вынести в `run(videos_per_page: int)`

BergaBruh committed:
update_films_data: Исправление ошибок

Добавлены недостающие type hints. Добавлено описание для класса и функции. Переделаны некоторые моменты по поводу вывода ошибок.

BergaBruh commented:
Скрипт был подправлен в некоторых местах. Требуется перепроверка кода.

ArthurO-micimp commented on content-api/python/update_films_data.py:
```diff
@@ -0,0 +1,72 @@
+        pages = count // limit
+        if count % limit != 0:
+            pages += 1
+        return pages+1
```
Всё таки тут я был не прав, +1 действительно должно быть внутри range для того чтобы последнюю страницу включать в список.

ArthurO-micimp commented on content-api/python/update_films_data.py:
```diff
@@ -0,0 +1,72 @@
+    def run(self):
+        """Запуск скрипта обновления метаданных фильмов и сериалов в Smarty.
+
+        Args:
+            videos_per_page (:obj:`int`): Количество фильмов и сериалов на одну страницу.
+            response (:obj:`SmartyContentAPI`): Ответ, получаемый от Smarty. В данном случаи - список видео.
+            pages (:obj:`int`): Номер страницы.
```
Всё нормально, только вот эту часть надо убрать. Аргументов у метода run нет, поэтому и описывать их не надо.

ArthurO-micimp reviewed (changes requested):
Небольшие фиксы внести и всё ок.

BergaBruh committed:
update_films_data: Финальный фикс

BergaBruh commented:
Сделал последние исправления.
